enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,441 @@
+```
+
+### Enable/disable the router
+The following configuration is proposed:
+```yaml
+ingress:
+  status: <Enabled|Disabled> # Defaults to Enabled.
+```
+
+With this option MicroShift will decide whether it should create the router
+upon starting. This includes not just the pod, but all the associated resources
+that come along with it: namespace, services, configmaps, etc.
+
+When setting the option to `Disabled`, the next MicroShift restart should delete
+all the router related resources, including APIs and controllers.
+
+Setting the option to `Enabled` will deploy the router, while `Disabled` will disable
+it. Default will be `Enabled`.
+
+### Listening ports
+The following configuration is proposed:
+```yaml
+ingress:
+  ports:
+    http: <int> # Defaults to 80.
+    https: <int> # Defaults to 443.
+```
+
+MicroShift does not own the host, which means there might be other
+applications running alongside and some ports may already be opened.
+MicroShift must be able to accommodate such situations and be ready to listen
+in other ports, hence the configuration options.
+
+In order to allow this configuration and many other advantages, the
+router service shall be changed so that it is exposed using `LoadBalancer`
+type instead of using host ports. See the Design Details section for more
+information.
+
+#### Firewalling ports
+Using `LoadBalancer` service type prevents the usage of firewalld to block
+access to ports. Since this is a feature that users may require for auditing,
+an alternative is provided.
+
+The `NetworkPolicy` resource fits the needs for blocking traffic to services
+and pods. Policies can also be configured to allow specific IP addresses and
+provide a high degree of customization for network traffic. They use the CNI
+instead of host level settings, providing greater flexibility and being self
+contained within MicroShift.
+
+Users in need of blocking traffic shall use the `NetworkPolicy` resource. A
+section in the documentation will be added to describe how to do this.
+
+### Configure IPs where the router listens
+The following configuration is proposed:
+```yaml
+ingress:
+  expose:
+    - <IP address/hostname> # Defaults to the node's hostname.
+```
+
+As we have seen in previous sections, the use of `LoadBalancer` makes ovnk
+configure iptables rules to expose the service outside of the cluster.
+Building on [LoadBalancer controller enhancement](loadbalancer-service-support.md),
+the service is exposed using the node IP and configured ports.
+
+All the IPs where the router is exposed need to be part of the `LoadBalancer`
+service status definition. In order to provide more flexibility, the
+configuration options are able to take IP addresses and/or hostnames.
+
+MicroShift will use the hostname as a default value for this. The router will
+only listen on the primary IP for the host unless configured to do otherwise.
+This is a sensible default that covers basic networking setups. Any advanced
```

Comment:
The new default behavior MUST maintain backwards compatibility. The default needs to be the nodeIP, not the "hostname", which may not be on the IP with the default route.

Reply:
+1 on Dougs comment. From my point of view, we need a way to express "listen on all available IPs" and make that default. The new implementation should default to the same behavior we had in the old, i.e. listen on all available IPs (without a need to specify them). could be e.g. "expose: *", or "expose: all", or make expose optional and if not present, it means "listen on all available IPs".

Reply:
Changed it so it listens on any IP, even the internal ones. The only one that does not work is the link local address, which is internal to ovnk and should not be used anywhere else. The current implementation listens there by mistake.